src/core/sorted_map.h

```diff
@@ -40,60 +49,250 @@ class SortedMap {
   static std::unique_ptr<SortedMap> FromListPack(const uint8_t* lp);
 
   size_t Size() const {
-    return zsl_->length;
+    return std::visit(Overload{[](const RdImpl& impl) { return impl.Size(); },
+                               [](const DfImpl& impl) { return impl.Size(); }},
+                      impl_);
   }
 
   bool Reserve(size_t sz) {
-    return dictExpand(dict_, sz) == DICT_OK;
+    return std::visit(Overload{[&](RdImpl& impl) { return impl.Reserve(sz); },
+                               [&](DfImpl& impl) { return impl.Reserve(sz); }},
+                      impl_);
   }
 
   // Interface equivalent to zsetAdd.
-  int Add(double score, sds ele, int in_flags, int* out_flags, double* newscore);
+  // Does not take ownership over ele string.
+  // Returns 1 if succeeded, false if the final score became invalid due to the update.
+  // newscore is set to the new score of the element only if in_flags contains ZADD_IN_INCR.
+  int Add(double score, sds ele, int in_flags, int* out_flags, double* newscore) {
+    return std::visit(
+        Overload{[&](RdImpl& impl) { return impl.Add(score, ele, in_flags, out_flags, newscore); },
+                 [&](DfImpl& impl) { return impl.Add(score, ele, in_flags, out_flags, newscore); }},
+        impl_);
+  }
```

Comment:
Didn't want to bother you refactoring, but either way... 100LOC less is good 😆 
Just a side note, you can use auto-lambdas (template under the hood), so you don't need to type it all out twice, i.e.

```c++
visit([](auto& impl) { impl.Add(); }, impl_)
```
```

Reply:
I was not sure I can, I thought this feature actually requires explicit override so it will deduce the base class types. checking it now

Reply:
indeed, it works. PTAL